docs/architecture/adr-042-state-sync.md

```diff
@@ -0,0 +1,209 @@
+and peers respond with some subset of the
+requested range and with notes on how to request the rest in parallel from other
+peers. Unlike chunk numbers, keys can be verified directly. And if some keys in the
+range are ommitted, proofs for the range will fail to verify.
+This way a node can start by requesting the entire tree from one peer,
+and that peer can respond with say the first few keys, and the ranges to request
+from other peers.
+
+#### Eager StateSync
+Warp Sync as implemented in Parity
+["Warp Sync"](https://wiki.parity.io/Warp-Sync-Snapshot-Format.html) to rapidly
+download both blocks and state snapshots from peers. Data is carved into ~4MB
+chunks and snappy compressed. Hashes of snappy compressed chunks are stored in a
+manifest file which co-ordinates the state-sync. Obtaining a correct manifest
+file seems to require an honest majority of peers. This means you may not find
+out the state is incorrect until you download the whole thing and compare it
+with a verified block header.
+
+A similar solution was implemented by Binance in
+[#3594](https://github.com/tendermint/tendermint/pull/3594)
+based on their initial implementation in
+[PR #3243](https://github.com/tendermint/tendermint/pull/3243)
+and [some learnings](https://docs.google.com/document/d/1npGTAa1qxe8EQZ1wG0a0Sip9t5oX2vYZNUDwr_LVRR4/edit).
+Note this still requires the honest majority peer assumption.
+One major advantage of the warp-sync approach is that chunks can be compressed before sending.
+
+Ideally State Sync in Tendermint would not require an honest majority of peers, and instead
+would rely on the standard light client security, meaning state chunks would be
+verifiable against the merkle root of the state. Such a design for tendermint
+was originally tracked in [#828](https://github.com/tendermint/tendermint/issues/828).
+
+### Analysis of Lazy vs Eager
+
+Lazy vs Eager have more in common than they differ. They all require
+reactors on the tendermint side, a set of ABCI messages and a method for
+serializing/deserializing snapshots facilitated by a SnapshotFormat.
+
+The biggest difference between Lazy and Eager proposals is in the
+read/write patterns necessitated by when serving a snapshot chunk.
+Specifically, Lazy State Sync random reads to the underlying data
+structure while Eager can optimize for sequential reads.
+
+This distinctin between approaches was demonstrated by Binance's [ackratos](https://github.com/ackratos) in their
+implementation of [Lazy State sync](https://github.com/tendermint/tendermint/pull/3243), The [analysis](https://docs.google.com/document/d/1npGTAa1qxe8EQZ1wG0a0Sip9t5oX2vYZNUDwr_LVRR4/) of the performance, and
+follow up implementation of [Warp Sync](http://github.com/tendermint/tendermint/pull/3594).
+
+One proposed difference between Lazy and Eager state sync is the safety
+model. It had been suggested that Lazy State Sync could use light client
```

Comment:
I am not sure to agree with the first sentence. You could use light client validation in both approaches.